File: maze_flutter/lib/models/maze.dart
class Point {
  final int x;
  final int y;
  const Point(this.x, this.y);

  Point operator +(Point other) => Point(x + other.x, y + other.y);

  @override
  bool operator ==(Object other) =>
      other is Point && other.x == x && other.y == y;

  @override
  int get hashCode => x.hashCode ^ y.hashCode;

  Map<String, int> toJson() => {'x': x, 'y': y};

  factory Point.fromJson(Map<String, dynamic> json) =>
      Point(json['x'] as int, json['y'] as int);
}

class Tile {
  static const int floor = 0;
  static const int wall = 1;
  static const int key = 2;
  static const int door = 3;
  static const int treasure = 4;
  static const int hidden = -1;
  static const int openDoor = 5;
  static const int start = 6;
}

class Maze {
  static const int width = 7;
  static const int height = 7;

  final List<List<int>> cells;

  Maze() : cells = List.generate(height, (_) => List.filled(width, Tile.floor)) {
    cells[0][0] = Tile.start; // default start position
  }

  Maze.from(List<List<int>> data)
      : cells = data.map((r) => List<int>.from(r)).toList();

  int get(int x, int y) {
    if (x < 0 || x >= width || y < 0 || y >= height) return -1;
    return cells[y][x];
  }

  void set(int x, int y, int tile) {
    if (x >= 0 && x < width && y >= 0 && y < height) {
      cells[y][x] = tile;
    }
  }

  int countTile(int tile) {
    int count = 0;
    for (final row in cells) {
      for (final c in row) {
        if (c == tile) count++;
      }
    }
    return count;
  }

  bool get hasKey => countTile(Tile.key) == 1;
  bool get hasDoor => countTile(Tile.door) == 1;
  bool get hasTreasure => countTile(Tile.treasure) == 1;
  bool get hasStart => countTile(Tile.start) == 1;

  Point get startPos => _findTile(Tile.start) ?? const Point(0, 0);

  bool get hasRequiredTiles =>
      hasKey && hasDoor && hasTreasure && hasStart && countTile(Tile.wall) <= 20;

  bool get isLocallyValid => hasRequiredTiles && isSolvable;

  /// BFS solvability: start→key (no door), key→door (no door blocking),
  /// door→treasure (door open).
  bool get isSolvable {
    if (!hasRequiredTiles) return false;
    final keyPos = _findTile(Tile.key);
    final doorPos = _findTile(Tile.door);
    final treasurePos = _findTile(Tile.treasure);
    if (keyPos == null || doorPos == null || treasurePos == null) return false;

    return _canReach(startPos, keyPos, doorOpen: false) &&
        _canReach(keyPos, doorPos, doorOpen: false) &&
        _canReach(doorPos, treasurePos, doorOpen: true);
  }

  /// Human-readable validation error, or null if valid.
  String? get validationError {
    if (!hasStart) return 'Place a start tile';
    if (!hasKey) return 'Place a key';
    if (!hasDoor) return 'Place a door';
    if (!hasTreasure) return 'Place a treasure';
    if (countTile(Tile.wall) > 20) return 'Too many walls (max 20)';

    final keyPos = _findTile(Tile.key);
    final doorPos = _findTile(Tile.door);
    final treasurePos = _findTile(Tile.treasure);

    if (!_canReach(startPos, keyPos!, doorOpen: false)) {
      return "Can't reach the key from start";
    }
    if (!_canReach(keyPos, doorPos!, doorOpen: false)) {
      return "Can't reach the door from key";
    }
    if (!_canReach(doorPos, treasurePos!, doorOpen: true)) {
      return "Can't reach the treasure from door";
    }
    return null;
  }

  Point? _findTile(int tile) {
    for (int y = 0; y < height; y++) {
      for (int x = 0; x < width; x++) {
        if (get(x, y) == tile) return Point(x, y);
      }
    }
    return null;
  }

  bool _canReach(Point from, Point to, {required bool doorOpen}) {
    if (from == to) return true;
    final visited = List.generate(height, (_) => List.filled(width, false));
    final queue = <Point>[from];
    visited[from.y][from.x] = true;
    const dirs = [Point(0, 1), Point(0, -1), Point(-1, 0), Point(1, 0)];

    while (queue.isNotEmpty) {
      final cur = queue.removeAt(0);
      for (final d in dirs) {
        final nx = cur.x + d.x, ny = cur.y + d.y;
        if (nx < 0 || nx >= width || ny < 0 || ny >= height) continue;
        if (visited[ny][nx]) continue;
        final next = Point(nx, ny);
        if (next == to) return true;
        final tile = get(nx, ny);
        if (tile == Tile.wall) continue;
        if (tile == Tile.door && !doorOpen) continue;
        visited[ny][nx] = true;
        queue.add(next);
      }
    }
    return false;
  }

  List<List<int>> toJson() => cells;

  Maze copy() => Maze.from(cells);
}

class SavedMaze {
  final String id;
  final String name;
  final List<List<int>> cells;
  final DateTime createdAt;

  SavedMaze({
    required this.id,
    required this.name,
    required this.cells,
    required this.createdAt,
  });

  factory SavedMaze.fromJson(Map<String, dynamic> json) => SavedMaze(
    id: json['id'] as String,
    name: json['name'] as String,
    cells: (json['cells'] as List).map((r) => List<int>.from(r as List)).toList(),
    createdAt: DateTime.parse(json['created_at'] as String),
  );

  Map<String, dynamic> toJson() => {
    'id': id,
    'name': name,
    'cells': cells,
    'created_at': createdAt.toIso8601String(),
  };

  Maze toMaze() => Maze.from(cells);
}

File: maze-server/internal/maze/maze.go
package maze

import "fmt"

// Tile codes (base)
const (
	TileFloor    = 0
	TileWall     = 1
	TileKey      = 2
	TileDoor     = 3
	TileTreasure = 4
	TileStart    = 6
)

// Discovered tile codes (runtime)
const (
	TileHidden  = -1
	TileOpenDoor = 5
)

// Direction constants
const (
	DirUp    = "up"
	DirDown  = "down"
	DirLeft  = "left"
	DirRight = "right"
)

// Move events
const (
	EventNone        = ""
	EventFoundKey    = "found_key"
	EventDoorLocked  = "door_locked"
	EventDoorOpened  = "door_opened"
	EventFoundTreasure = "found_treasure"
	EventHitWall     = "hit_wall"
)

// Point represents a canonical grid position.
// x: 0..width-1 left to right, y: 0..height-1 bottom to top.
type Point struct {
	X int `json:"x"`
	Y int `json:"y"`
}

// Maze represents a player-built maze.
type Maze struct {
	Width  int   `json:"width"`
	Height int   `json:"height"`
	Cells  [][]int `json:"cells"`
}

// NewMaze creates an empty maze filled with floor tiles.
func NewMaze(w, h int) *Maze {
	cells := make([][]int, h)
	for y := range cells {
		cells[y] = make([]int, w)
	}
	return &Maze{Width: w, Height: h, Cells: cells}
}

// Get returns the tile at canonical (x, y).
func (m *Maze) Get(x, y int) int {
	if x < 0 || x >= m.Width || y < 0 || y >= m.Height {
		return -1
	}
	return m.Cells[y][x]
}

// Set sets the tile at canonical (x, y).
func (m *Maze) Set(x, y, tile int) {
	if x >= 0 && x < m.Width && y >= 0 && y < m.Height {
		m.Cells[y][x] = tile
	}
}

// StartPos returns the start position by finding the start tile, or (0,0) as fallback.
func (m *Maze) StartPos() Point {
	for y := 0; y < m.Height; y++ {
		for x := 0; x < m.Width; x++ {
			if m.Get(x, y) == TileStart {
				return Point{X: x, Y: y}
			}
		}
	}
	return Point{X: 0, Y: 0}
}

// Validate checks that the maze is valid per MVP rules.
// Returns nil if valid, an error describing the problem otherwise.
func (m *Maze) Validate() error {
	if m.Width != 7 || m.Height != 7 {
		return fmt.Errorf("maze must be 7x7, got %dx%d", m.Width, m.Height)
	}

	// Count special tiles
	var keyPos, doorPos, treasurePos, startPos *Point
	wallCount := 0
	for y := 0; y < m.Height; y++ {
		for x := 0; x < m.Width; x++ {
			switch m.Get(x, y) {
			case TileStart:
				if startPos != nil {
					return fmt.Errorf("multiple start tiles found")
				}
				p := Point{X: x, Y: y}
				startPos = &p
			case TileKey:
				if keyPos != nil {
					return fmt.Errorf("multiple keys found")
				}
				p := Point{X: x, Y: y}
				keyPos = &p
			case TileDoor:
				if doorPos != nil {
					return fmt.Errorf("multiple doors found")
				}
				p := Point{X: x, Y: y}
				doorPos = &p
			case TileTreasure:
				if treasurePos != nil {
					return fmt.Errorf("multiple treasures found")
				}
				p := Point{X: x, Y: y}
				treasurePos = &p
			case TileWall:
				wallCount++
			}
		}
	}

	if startPos == nil {
		return fmt.Errorf("maze must have exactly one start tile")
	}
	if keyPos == nil {
		return fmt.Errorf("maze must have exactly one key")
	}
	if doorPos == nil {
		return fmt.Errorf("maze must have exactly one door")
	}
	if treasurePos == nil {
		return fmt.Errorf("maze must have exactly one treasure")
	}
	if wallCount > 20 {
		return fmt.Errorf("too many walls: %d (max 20)", wallCount)
	}

	// Reachability: start -> key (no door crossing)
	if !m.reachable(*startPos, *keyPos, false) {
		return fmt.Errorf("key is not reachable from start without crossing door")
	}

	// Reachability: key -> door (no door blocking, since we just need to reach the door cell)
	if !m.reachable(*keyPos, *doorPos, false) {
		return fmt.Errorf("door is not reachable from key")
	}

	// Reachability: door -> treasure (door is now open)
	if !m.reachable(*doorPos, *treasurePos, true) {
		return fmt.Errorf("treasure is not reachable from door")
	}

	// Minimum path length check
	pathLen := m.shortestPath(*startPos, *keyPos, false) +
		m.shortestPath(*keyPos, *doorPos, false) +
		m.shortestPath(*doorPos, *treasurePos, true)
	if pathLen < 8 {
		return fmt.Errorf("maze is too easy: shortest solution is %d moves (min 8)", pathLen)
	}

	return nil
}

// reachable checks if dst is reachable from src using BFS.
// If doorOpen is false, door tiles block passage.
func (m *Maze) reachable(src, dst Point, doorOpen bool) bool {
	return m.shortestPath(src, dst, doorOpen) >= 0
}

// shortestPath returns the shortest path length from src to dst, or -1 if unreachable.
func (m *Maze) shortestPath(src, dst Point, doorOpen bool) int {
	if src == dst {
		return 0
	}

	visited := make([][]bool, m.Height)
	for i := range visited {
		visited[i] = make([]bool, m.Width)
	}

	type state struct {
		pos  Point
		dist int
	}

	queue := []state{{pos: src, dist: 0}}
	visited[src.Y][src.X] = true

	dirs := []Point{{0, 1}, {0, -1}, {-1, 0}, {1, 0}}

	for len(queue) > 0 {
		cur := queue[0]
		queue = queue[1:]

		for _, d := range dirs {
			nx, ny := cur.pos.X+d.X, cur.pos.Y+d.Y
			if nx < 0 || nx >= m.Width || ny < 0 || ny >= m.Height {
				continue
			}
			if visited[ny][nx] {
				continue
			}

			next := Point{X: nx, Y: ny}

			// Check destination before applying tile filter —
			// we can always reach the destination tile itself.
			if next == dst {
				return cur.dist + 1
			}

			tile := m.Get(nx, ny)
			if tile == TileWall {
				continue
			}
			if tile == TileDoor && !doorOpen {
				continue
			}

			visited[ny][nx] = true
			queue = append(queue, state{pos: next, dist: cur.dist + 1})
		}
	}

	return -1
}
